src/can.rs

```diff
@@ -0,0 +1,125 @@
+//! # Controller Area Network (CAN) Interface
+//!
+//!
+//! TX: Alternate Push-Pull Output
+//! RX: Input Floating Input
+//!
+//! ### CAN1
+//!
+//! | Function | Pins  |
+//! |----------|-------|
+//! | TX       | PB9   |
+//! | RX       | PB8   |
+//!
+//! ### CAN2
+//!
+//! | Function | Pins  |
+//! |----------|-------|
+//! | TX       | PB13  |
+//! | RX       | PB12  |
+
+use crate::bb;
+use crate::gpio::{
+    gpioa::{PA11, PA12},
+    gpiob::{PB12, PB13, PB5, PB6, PB8, PB9},
+    Alternate, AF9,
+};
+use crate::pac::{CAN1, CAN2};
+use crate::rcc::APB1;
+
+mod sealed {
+    pub trait Sealed {}
+
+    /// Bus associated to peripheral
+    pub trait RccBus {
+        /// Bus type;
+        type Bus;
+    }
+}
+
+pub trait Pins: sealed::Sealed {
+    type Instance;
+}
+
+/*
+    order: tx, rx similar to serial
+*/
+
+macro_rules! pins {
+    ($($PER:ident => ($tx:ident, $rx:ident),)+) => {
+        $(
+            impl sealed::Sealed for ($tx<Alternate<AF9>>, $rx<Alternate<AF9>>) {}
+            impl Pins for ($tx<Alternate<AF9>>, $rx<Alternate<AF9>>) {
+                type Instance = $PER;
+            }
+        )+
+    }
+}
+
+pins! {
+    CAN1 => (PA12, PA11),
+    CAN1 => (PB9, PB8),
+    CAN2 => (PB13, PB12),
+    CAN2 => (PB6, PB5),
+}
+
+use sealed::RccBus;
+
+/// Enable/disable peripheral
+pub trait Enable: RccBus {
+    fn enable(apb: &mut Self::Bus);
+}
+
+macro_rules! bus {
+    ($($PER:ident => ($apbX:ty, $peren:literal),)+) => {
+        $(
+            impl RccBus for crate::pac::$PER {
+                type Bus = $apbX;
+            }
+            impl Enable for crate::pac::$PER {
+                #[inline(always)]
+                fn enable(apb: &mut Self::Bus) {
+                    unsafe { bb::set(apb.enr(), $peren) };
+                }
+            }
+        )+
+    }
+}
+
+bus! {
+    CAN1 => (APB1, 25),
+    CAN2 => (APB1, 26),
+}
+
+/// Interface to the CAN peripheral.
+pub struct Can<Instance> {
+    _peripheral: Instance,
+}
+
+impl<Instance> Can<Instance>
+where
+    Instance: Enable<Bus = APB1>,
+{
+    /// Creates a CAN interaface.
+    pub fn new<P>(can: Instance, _pins: P, apb: &mut APB1) -> Can<Instance>
```

Comment:
Its a common pattern to take ownership of the pins.
For CAN it might be a bit too restrictive though: https://github.com/stm32-rs/stm32f1xx-hal/pull/215 (section *Separate assign_pins()*)